telegram/ext/filters.py

```diff
@@ -850,29 +851,68 @@ class user(BaseFilter):
         Examples:
             ``MessageHandler(Filters.user(1234), callback_method)``
 
+        Attributes:
+            user_ids(set(:obj:`int`), optional): Which user ID(s) to allow through.
+            usernames(set(:obj:`str`), optional): Which username(s) (without leading '@') to allow
+                through.
         Args:
-            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow through.
-            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow through.
-                If username starts with '@' symbol, it will be ignored.
+            user_id(:obj:`int` | List[:obj:`int`], optional): Which user ID(s) to allow
+                through.
+            username(:obj:`str` | List[:obj:`str`], optional): Which username(s) to allow
+                through. If username starts with '@' symbol, it will be ignored.
 
         Raises:
             ValueError: If chat_id and username are both present, or neither is.
 
         """
 
         def __init__(self, user_id=None, username=None):
-            if not (bool(user_id) ^ bool(username)):
+            if (user_id is None) == (username is None):
                 raise ValueError('One and only one of user_id or username must be used')
-            if user_id is not None and isinstance(user_id, int):
-                self.user_ids = [user_id]
-            else:
-                self.user_ids = user_id
-            if username is None:
-                self.usernames = username
-            elif isinstance(username, string_types):
-                self.usernames = [username.replace('@', '')]
-            else:
-                self.usernames = [user.replace('@', '') for user in username]
+
+            self._user_ids_lock = Lock()
+            self._usernames_lock = Lock()
+
+            # Initialize in a way that will not fail the first setter calls
+            self._user_ids = user_id
+            self._usernames = username
+            # Actually initialize
+            self.user_ids = user_id
+            self.usernames = username
+
+        @property
+        def user_ids(self):
+            with self._user_ids_lock:
+                return self._user_ids
+
+        @user_ids.setter
+        def user_ids(self, user_id):
+            if (user_id is None) == (self.usernames is None):
+                raise ValueError('One and only one of user_id or username must be used')
```

Comment:
1. Take the lock before looking at privates, not just before you're about to write to them.
2. Thinking of Locks, you may want to consider a RwLock, but for the simplicity of first implementation lets stick to a plain Lock.
3. Since I'm considering an empty Set and None to be equivalent, code may be simplified a bit:
```suggestion
          with self.__lock:
              if user_id and self.usernames:
                    raise RuntimeError("Can't set user_id in conjunction with (already set) usernames")
```